File: 20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
from collections import deque


class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2:
            return False
        pairmp = {"(": ")", "[": "]", "{": "}"}
        stack = deque()

        for c in s:
            if c in pairmp:
                stack.append(c)
            elif not stack or pairmp[stack.pop()] != c:
                return False
        return not stack


# @lc code=end

File: 20.valid-parentheses.rs
/*
 * @lc app=leetcode id=20 lang=rust
 *
 * [20] Valid Parentheses
 */

// @lc code=start
use std::collections::VecDeque;
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack = VecDeque::new();
        for c in s.chars() {
            match c {
                '{' => stack.push_back('}'),
                '(' => stack.push_back(')'),
                '[' => stack.push_back(']'),
                '}' | ')' | ']' if Some(c) != stack.pop_back() => return false,
                _ => (),
            }
        }
        stack.is_empty()
    }
}
// @lc code=end

